### ISSUE TITLE
Restic tries to create lock file on forget even with `--no-lock` and `--dry-run`

### ISSUE THREAD

**Issue by sza-1** [restic/restic#3464]
Output of `restic version`
--------------------------
restic 0.12.0 compiled with go1.16rc1 on linux/amd64

How did you run restic exactly?
-------------------------------
`restic --password-command 'printf test' --repo test --no-lock forget --dry-run --prune latest`

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
rclone (and local).

Expected behavior
-----------------
restic should permit to forget/prune snapshots when `--no-lock` specified

Actual behavior
---------------
restic failed to forget/prune snapshot when repo does not have space for lock file even `--no-lock` specified.

```
Save(<lock/d400f5ba6a>) returned error, retrying after 873.42004ms: server response unexpected: 500 Internal Server Error (500)
rclone: 2021/07/25 00:52:07 ERROR : locks/d400f5ba6ab83e25e62036dd59d23e924a6395b4242d2130a503114e4c66edfc: Post request put error: failed to get upload URL: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap. (403 storage_cap_exceeded)
```
or

```
Save(<lock/f266849aaa>) returned error, retrying after 552.330144ms: OpenFile: open test/locks/f266849aaa262ea100c69c6859039629088c6038f9f21e6287de721312807324: operation not permitted
```

Steps to reproduce the behavior
-------------------------------
- Create test repo.
- Backup some files.
- Ensure that there is no free space for lock file or use `chattr +i locks` on Linux: `sudo chattr +i test/locks`
- Run `restic forget`:

```
$  restic --password-command 'printf test' --repo test --no-lock forget --dry-run --prune latest
repository 02692054 opened successfully, password is correct
Save(<lock/b3d6757227>) returned error, retrying after 552.330144ms: OpenFile: open test/locks/b3d67572272f43ddb12c1710ea85ad2fa85a3798e72b330df9f6b824d9cac9b9: operation not permitted
...
open test/locks/b3d67572272f43ddb12c1710ea85ad2fa85a3798e72b330df9f6b824d9cac9b9: operation not permitted
OpenFile
github.com/restic/restic/internal/backend/local.(*Local).Save
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/backend/local/local.go:117
github.com/restic/restic/internal/limiter.rateLimitedBackend.Save
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/limiter/limiter_backend.go:30
github.com/restic/restic/internal/backend.(*RetryBackend).Save.func1
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/backend/backend_retry.go:66
github.com/cenkalti/backoff.RetryNotifyWithTimer
	/usr/share/gocode/src/github.com/cenkalti/backoff/retry.go:55
github.com/cenkalti/backoff.RetryNotify
	/usr/share/gocode/src/github.com/cenkalti/backoff/retry.go:34
github.com/restic/restic/internal/backend.(*RetryBackend).retry
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/backend/backend_retry.go:46
github.com/restic/restic/internal/backend.(*RetryBackend).Save
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/backend/backend_retry.go:60
github.com/restic/restic/internal/cache.(*Backend).Save
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/cache/backend.go:54
github.com/restic/restic/internal/repository.(*Repository).SaveUnpacked
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/repository/repository.go:319
github.com/restic/restic/internal/repository.(*Repository).SaveJSONUnpacked
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/repository/repository.go:299
github.com/restic/restic/internal/restic.(*Lock).createLock
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/restic/lock.go:163
github.com/restic/restic/internal/restic.newLock
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/restic/lock.go:107
github.com/restic/restic/internal/restic.NewExclusiveLock
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/internal/restic/lock.go:75
main.lockRepository
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/cmd/restic/lock.go:35
main.lockRepoExclusive
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/cmd/restic/lock.go:26
main.runForget
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/cmd/restic/cmd_forget.go:101
main.glob..func11
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/cmd/restic/cmd_forget.go:28
github.com/spf13/cobra.(*Command).execute
	/usr/share/gocode/src/github.com/spf13/cobra/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/usr/share/gocode/src/github.com/spf13/cobra/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/usr/share/gocode/src/github.com/spf13/cobra/command.go:895
main.main
	/builddir/build/BUILD/restic-0.12.0/_build/src/github.com/restic/restic/cmd/restic/main.go:98
runtime.main
	/usr/lib/golang/src/runtime/proc.go:225
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1371
unable to create lock in backend
```

Do you have any idea what may have caused this?
-----------------------------------------------
main.runForget calls lockRepoExclusive in https://github.com/restic/restic/blob/v0.12.0/cmd/restic/cmd_forget.go#L101

```
	lock, err := lockRepoExclusive(gopts.ctx, repo)
	defer unlockRepo(lock)
	if err != nil {
		return err
	}
```

Do you have an idea how to solve the issue?
-------------------------------------------
Probably main.runForget in https://github.com/restic/restic/blob/v0.12.0/cmd/restic/cmd_forget.go#L101 should not call lockRepoExclusive when `--no-lock` specified. At least it should not call lockRepoExclusive when `--dry-run` specified.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Someday the bugs will run out.

**Comment by MichaelEischer** [comment]
`prune` won't work if there is no free space at all, see #1153. In order to prevent repository corruption it is mandatory for `prune` to lock the repository, to forbid concurrent modifications.

I agree that `forget` could avoid locking the repository if `--dry-run` was specified. Holding an exclusive lock while running `forget` (without the `--prune` part!) might not be strictly necessary. But as `prune` will require such a lock anyways, there's not much to gain there.

**Comment by crazyh** [comment]
> `prune` won't work if there is no free space at all, see #1153. In order to prevent repository corruption it is mandatory for `prune` to lock the repository, to forbid concurrent modifications.
> 
> I agree that `forget` could avoid locking the repository if `--dry-run` was specified. Holding an exclusive lock while running `forget` (without the `--prune` part!) might not be strictly necessary. But as `prune` will require such a lock anyways, there's not much to gain there.

Ahh, got it. Thanks for the explanation.

TL;DR

Just faced with a nasty case: accidentally backed up huge libvirt images to repo, ate all limits on B2 and now even can not remove the huge snapshots without asking to remove caps for the account. :)

Even though https://restic.readthedocs.io/en/latest/060_forget.html and #1153 describe forget/prune internals, probably some part of the documentation should state a bold warning about free space requirements for the prune operation.

**Comment by rawtaz** [comment]
> probably some part of the documentation should state a bold warning about free space requirements for the prune operation.

Can you suggest something that you think is both generic enough and that would have worked for you to avoid the unexpected?

**Comment by crazyh** [comment]
> > probably some part of the documentation should state a bold warning about free space requirements for the prune operation.
> 
> Can you suggest something that you think is both generic enough and that would have worked for you to avoid the unexpected?

Probably it should say (in the beginning of `Removing backup snapshots` section or in the begining of man page) something like the following:

```
WARNING: manual intervention may be required if repo is running out of free space.
```

```
(You can not cleanup the repo with restic if repo does not have enough free space).
```

Or even the following:

```
Please make sure there is enough free space for the prune operation to work.

See `Customize pruning` in `Removing backup snapshots` section of restic documentation for details about how the pruning process works.

Also keep in mind that the `--no-lock` flag does not affect all commands even when `--dry-run` is enabled. Lock file creation requires free space too.
```

**Comment by MichaelEischer** [comment]
@crazyh I guess the backup run which backed up the libvirt images didn't finish? Then the easiest way would be to temporarily comment out the `lockRepoExclusive` step in the `prune` command. Then `restic prune --max-repack-size 0` should be able to free enough space to allow prune to complete.

**Comment by crazyh** [comment]
> @crazyh I guess the backup run which backed up the libvirt images didn't finish or at least some of them? Then the easiest way would be to temporarily comment out the `lockRepoExclusive` step in the `prune` command. Then `restic prune --max-repack-size 0` should be able to free enough space to allow prune to complete.

Yes, as I remember it did not finish. I solved it by changing caps in B2. But thanks for knowledge about `--max-repack-size` - for me it was not obvious and can help in case of future issues.

**Comment by MichaelEischer** [comment]
Prune removes unreferenced pack files as one of the early steps. That should be able to free up enough space to allow prune to update the repository index. `--max-repack-size` now tells prune to not repack anything, but to only delete pack files which are no used at all. Thus the index update will also allow prune to delete these pack files. At that point, there should be enough free space to continue normal operations.

So what's left to do for this issue (reducing the amount of free space necessary for prune to work belongs into #1153) would be to allow `forget` to avoid locking the repository if `--dry-run` was specified. Is anyone up for a PR?

**Comment by tleydxdy** [comment]
I was trying to do a dry run with `restic forget --prune -l 2 --dry-run` and was quite surprised that restic seems to actually try to prune the repo, this doesn't feel intuitive at all. (is it only pretending to prune?)

**Comment by MichaelEischer** [comment]
> I was trying to do a dry run with `restic forget --prune -l 2 --dry-run` and was quite surprised that restic seems to actually try to prune the repo, this doesn't feel intuitive at all. (is it only pretending to prune?)

restic only calculates how much data prune would free, but does not delete anything. In my opinion that's actually the expected behavior, after all if one just wanted to know which snapshots will be deleted `forget --dry-run` (without `--prune`) is enough. However, as that question is off-topic here, please open a thread in the forum.restic.net if you want to discuss it further.

**Comment by MichaelEischer** [comment]
Fixed by #3464.
